src/popup-page/context/global.tsx

```diff
@@ -0,0 +1,53 @@
+import React, { createContext, useReducer } from 'react'
+
+export class GlobalStateActionError extends Error {
+  constructor(action: string, message: string) {
+    const formattedMessage = `Error happened on action "${action}": ${message} `
+    super(formattedMessage)
+  }
+}
+
+interface IState {
+  beeApiUrl: string
+}
+
+interface IAction<T = string> {
+  type: 'BEE_API_URL_CHANGE'
+  newValue?: T
+}
+
+interface ContextValue {
+  state: IState
+  dispatch: React.Dispatch<IAction<string>>
+}
+
+const initialState: IState = {
+  beeApiUrl: 'http://localhost:1633',
+}
+const GlobalContext = createContext<ContextValue>({
+  state: initialState,
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  dispatch: () => {},
+})
+const { Provider } = GlobalContext
+
+const GlobalStateProvider = ({ children }: { children: React.ReactElement }): React.ReactElement => {
+  const [state, dispatch] = useReducer((state: IState, action: IAction): IState => {
```

Comment:
The global state handler shouldn't be the only SSOT for shared states, but structurally what a UI state handler consists of and how should it handle data, you are right.
But what would be the workflow in your proposal for external local store changing? Because in this case not only the React App can change the store that the state manager handles.

Reply:
If there is no single source of truth, then as the application grows the data will be out of sync eventually. Here the idea is that every state change goes through the `GlobalContext`, even if it's not coming from the UI.

For example updating the `beeApiUrl` would be like this:

```typescript
useContext(GlobalContext).dispatch({ type: 'BEE_API_URL_CHANGE', newValue: '<new value>' })
```

Reply:
It can be out of sync indeed, but the background script and the script of popup page don't share common scope as I noticed. So the scripts can use common TS source files (like in the case of `src/utils/storage.ts`), but they cannot use common state directly, they communicate only via messages. In this case the local store change event message is perfect to fulfill this role.
One of my subsequent commit I wrapped the original `dispatch` function to alter localstorage when it is necessary.